### PR TITLE
Added BUILD flags to provision for building the kernel with non-upstream patches

### PR DESCRIPTION
**Pull request by vivekrnv**
Signed-off-by: Vivek Reddy <vkarri@nvidia.com>

#### Why I did it

Provide a way to include non-upstream kernel patches in the kernel build process

Requires https://github.com/sonic-net/sonic-linux-kernel/pull/296 and can be merged in any order

#### How I did it

Introduced two new environment variables:
```
EXTERNAL_KERNEL_PATCH_URL = <url to the archive containing patches>, empty by default
INCLUDE_EXTERNAL_PATCH_TAR= <y|n to include the patches under platform/ folder> 'n' by default
```

Note: For `INCLUDE_EXTERNAL_PATCH_TAR` to work, the platform has to override `EXTERNAL_KERNEL_PATCH_TAR `variable. And the variable should be a path pointing to the tarball containing patches

Use SONIC_OVERRIDE_BUILD_VARS to fuse in these env variables:

```
- NOJESSIE=1 NOSTRETCH=1 NOBUSTER=1 make SONIC_OVERRIDE_BUILD_VARS=' INCLUDE_EXTERNAL_PATCH_TAR=y ' target/debs/bullseye/linux-headers-5.10.0-12-2-common_5.10.103-1_all.deb
- NOJESSIE=1 NOSTRETCH=1 NOBUSTER=1 make SONIC_OVERRIDE_BUILD_VARS=' EXTERNAL_KERNEL_PATCH_URL=http://<url-root>/loc/patches.tar.gz ' target/debs/bullseye/linux-headers-5.10.0-12-2-common_5.10.103-1_all.deb
```

#### Structure of the patch tar archive
1) It should contain a file named series. series should provide an order of which the patches have to be applied
2) All the patches should be present in the archive in the same folder where series resides.
3) Developers should make sure patches apply cleanly over the existing patches present in the sonic-linux-kernel/ repo. 
```
admin@build-server:/sonic-buildimage/platform/mellanox/non-upstream-patches$ tar -tf patches.tar.gz
./
./mlx5-Implement-get_module_eeprom_by_page.patch
./mlx5-Add-support-for-DSFP-module-EEPROM-dumps.patch
./series
./mlx5-Refactor-module-EEPROM-query.patch
```

#### How to verify it

1) Build the kernel
2) Check if the patches are applied on the kernel source
```
vkarri@build-server:/sonic-buildimage/src/sonic-linux-kernel/linux-5.10.103$ git log --oneline -n 4
a524105fc (HEAD -> master, refs/patches/master/mlx5-Add-support-for-DSFP-module-EEPROM-dumps.patch) From 4c88fa412a100f925b8ab1aa952a672895f69d35 Mon Sep 17 00:00:00 2001 Subject: net/mlx5: Add support for DSFP module EEPROM dumps

8f366b27c (refs/patches/master/mlx5-Implement-get_module_eeprom_by_page.patch) From e109d2b204daa223e6d3cdaa369071c3ea96dcbf Mon Sep 17 00:00:00 2001 Subject: net/mlx5: Implement get_module_eeprom_by_page()

eb9bddfad (refs/patches/master/mlx5-Refactor-module-EEPROM-query.patch) From e19b0a3474ab9ef90dd110af9f39fc87329755f1 Mon Sep 17 00:00:00 2001 Subject: net/mlx5: Refactor module EEPROM query

cf79fff3a (refs/patches/master/armhf_secondary_boot_online.patch) From 500d48948f43d18a07894b5fd4f891afbade1d79 Mon Sep 17 00:00:00 2001
```

#### Which release branch to backport (provide reason below if selected)

- [ ] 201811
- [ ] 201911
- [ ] 202006
- [ ] 202012
- [ ] 202106
- [ ] 202111
- [ ] 202205

### CODE REVIEW

File: platform/mellanox/non-upstream-patches/README.md [hunk 1]
```diff
@@ -0,0 +1,25 @@
+## Mellanox non-upstream linux kernel patches ##
+
+To include non-upstream patches into the sonic-linux image during build time, this folder must contain a patch archive.
+
+### Structure of the patch archive
+
+  1. The tarball should be gzip compressed i.e. of extension .tar.gz
```
Comment (paulmenzel) [inline platform/mellanox/non-upstream-patches/README.md:7]: Why hard code the compression algorithm?

File: platform/mellanox/non-upstream-patches/README.md [hunk 2]
```diff
@@ -0,0 +1,25 @@
+## Mellanox non-upstream linux kernel patches ##
+
+To include non-upstream patches into the sonic-linux image during build time, this folder must contain a patch archive.
+
+### Structure of the patch archive
+
+  1. The tarball should be gzip compressed i.e. of extension .tar.gz
+  2. It should contain a file named series. series should provide an order of which the patches have to be applied
```
Comment (paulmenzel) [inline platform/mellanox/non-upstream-patches/README.md:8]: Please elaborate, how to create such a series file.

File: platform/mellanox/non-upstream-patches/README.md [hunk 3]
```diff
@@ -0,0 +1,25 @@
+## Mellanox non-upstream linux kernel patches ##
+
+To include non-upstream patches into the sonic-linux image during build time, this folder must contain a patch archive.
+
+### Structure of the patch archive
+
+  1. The tarball should be gzip compressed i.e. of extension .tar.gz
+  2. It should contain a file named series. series should provide an order of which the patches have to be applied
+  3. All the patches should be present in the archive in the same folder where series resides.
+  4. Developers should make sure patches apply cleanly over the existing patches present in the sonic-linux-kernel/ repo.
```
Comment (paulmenzel) [inline platform/mellanox/non-upstream-patches/README.md:10]: Add an URL to the archive/repository.

File: platform/mellanox/non-upstream-patches/README.md [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+### Structure of the patch archive
+
+  1. The tarball should be gzip compressed i.e. of extension .tar.gz
+  2. It should contain a file named series. series should provide an order of which the patches have to be applied
+  3. All the patches should be present in the archive in the same folder where series resides.
+  4. Developers should make sure patches apply cleanly over the existing patches present in the sonic-linux-kernel/ repo.
+
+#### Example
+```
+admin@build-server:/sonic-buildimage/platform/mellanox/non-upstream-patches$ tar -tvf patches.tar.gz
```
Comment (paulmenzel) [inline platform/mellanox/non-upstream-patches/README.md:14]: The verbose output with `-v` is not needed. Or are the permissions and ownership relevant? Then leave it.

File: platform/mellanox/rules.mk [hunk 5]
```diff
@@ -51,4 +51,9 @@ $(DOCKER_PLATFORM_MONITOR)_DEPENDS += $(APPLIBS) $(SX_COMPLIB) $(SXD_LIBS) $(SX_
 # Force the target bootloader for mellanox platforms to grub regardless of arch
 TARGET_BOOTLOADER = grub
 
+# Set y to include non upstream patches for mellanox
+INCLUDE_MLNX_PATCHES = n
```
Comment (paulmenzel) [inline platform/mellanox/rules.mk:55]: Why is this needed? Can’t patches for other subsystems be included?

**Review by paulmenzel — COMMENTED** [review]
Please use imperative mood in git commit messages summaries/titles, and merge/pull request summaries/titles. s/Added/Add/

File: platform/mellanox/non-upstream-patches/README.md (hunk 1 above)
Comment (saiarcot895) [reply platform/mellanox/non-upstream-patches/README.md:7]: For `tar`, the compression algorithm doesn't need to be specified to be able to extract the patches. I would instead rephrase this to specify various compression formats (.tar.gz, .tar.xz, .tar.bz2, etc.).

File: platform/mellanox/non-upstream-patches/README.md (hunk 3 above)
Comment (saiarcot895) [reply platform/mellanox/non-upstream-patches/README.md:10]: This is referring to the `src/sonic-linux-kernel` submodule. Perhaps specifying the path like that would be better for clarity.

File: slave.mk [hunk 6]
```diff
@@ -84,6 +85,7 @@ export MULTIARCH_QEMU_ENVIRON
 export DOCKER_BASE_ARCH
 export CROSS_BUILD_ENVIRON
 export BLDENV
+export BUILD_WORKDIR
```
Comment (saiarcot895) [inline slave.mk:88]: Is there a reason for `BUILD_WORKDIR` to get exported?

Comment (vivekrnv) [reply slave.mk:88]: Yes, It is required to make MLNX_PATCH_LOC  an absolute path `MLNX_PATCH_LOC = $(BUILD_WORKDIR)/$(PLATFORM_PATH)/non-upstream-patches/patches.tar.gz`

File: platform/mellanox/rules.mk (hunk 5 above)
Comment (vivekrnv) [reply platform/mellanox/rules.mk:55]: Other subsystems planning to include no-upstream patches can use EXTERNAL_KERNEL_PATCHES  flag to provide a public accessible link to the tarball

**Comment by vivekrnv** [comment]
@saiarcot895, I've updated the pr according to the comments given https://github.com/sonic-net/sonic-linux-kernel/pull/296. Can you also review this?

File: platform/mellanox/rules.mk [hunk 7]
```diff
@@ -51,4 +51,7 @@ $(DOCKER_PLATFORM_MONITOR)_DEPENDS += $(APPLIBS) $(SX_COMPLIB) $(SXD_LIBS) $(SX_
 # Force the target bootloader for mellanox platforms to grub regardless of arch
 TARGET_BOOTLOADER = grub
 
+# location for the platform specific external kernel patches tarball
+override EXTERNAL_KERNEL_PATCH_TAR := $(BUILD_WORKDIR)/$(PLATFORM_PATH)/non-upstream-patches/patches.tar.gz
```
Comment (saiarcot895) [inline platform/mellanox/rules.mk:55]: Is the `override` keyword needed here, or can a regular assignment just work?

Comment (vivekrnv) [reply platform/mellanox/rules.mk:55]: I guess, `EXTERNAL_KERNEL_PATCH_TAR = $(BUILD_WORKDIR)/$(PLATFORM_PATH)/non-upstream-patches/patches.tar.gz` this'll work as well. But i thought, override here would make it clear that this is the final value. I can change it to a regular assignment if you'd prefer

Comment (saiarcot895) [reply platform/mellanox/rules.mk:55]: Considering it's in a platform-specific file, I'm fine with it with an `override`. If needed, it can be removed later.

**Review by saiarcot895 — APPROVED** [review]
